saml2/saml2-service-provider/src/opensaml4Main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java

```diff
@@ -685,13 +748,44 @@ private static Saml2AuthenticationException createAuthenticationException(String
 		};
 	}
 
+	private static boolean assertionContainsInResponseTo(Assertion assertion) {
+		Subject subject = (assertion != null) ? assertion.getSubject() : null;
+		List<SubjectConfirmation> confirmations = (subject != null) ? subject.getSubjectConfirmations()
+				: new ArrayList<>();
+		return confirmations.stream().filter((confirmation) -> {
+			SubjectConfirmationData confirmationData = confirmation.getSubjectConfirmationData();
+			return confirmationData != null && StringUtils.hasText(confirmationData.getInResponseTo());
+		}).findFirst().orElse(null) != null;
+	}
+
+	private static void addRequestIdToValidationContext(AbstractSaml2AuthenticationRequest storedRequest,
+			Map<String, Object> context) {
+		String requestId = null;
+		try {
+			AuthnRequest request = parseRequest(storedRequest);
+			requestId = (request != null) ? request.getID() : null;
+		}
+		catch (Exception ex) {
+		}
+		if (StringUtils.hasText(requestId)) {
+			context.put(SAML2AssertionValidationParameters.SC_VALID_IN_RESPONSE_TO, requestId);
+		}
+		else {
+			context.put(SAML2AssertionValidationParameters.SC_VALID_IN_RESPONSE_TO, "");
```

Comment:
Here we set the empty string so that we know that the validation will fail (since the very prerequisite for even doing the validation is that the assertion has a non-empty `InResponseTo`). The reason for failing is either that there was no stored request, it could not be deserialized or it lacked id. You think this solution is ok?

Reply:
Could trigger a false positive if we have multiple confirmations and one has the (invalid) empty string as `InResponseTo` whereas a later one has a proper one ...

Reply:
Perhaps better to stay with the override anyways? 
* Won't be more code than is added by this solution
* Easier to understand what's going on
* Will cover all cases since we can control directly whether to consider valid or not

Reply:
I think you can just leave it unset. The OpenSAML code works as follows:

- If no `InResponseTo` in subject confirmation -> valid
- Or, if no `SC_VALID_IN_RESPONSE_TO` set -> invalid
- Or, if `SC_VALID_IN_RESPONSE_TO` mismatches `InResponseTo` -> invalid

> Could trigger a false positive

If any assertions are wrong, then we should fail. We don't accept a response that is in response to multiple requests.

> override anyways

Ideally, we don't write code that we don't have to as this lowers maintenance. I don't see why it's easier to understand what is going on. The code that you have written is virtually the same as OpenSAML's `BearerSubjectConfirmationValidator#validateInResponseTo`.

Reply:
Yeah of course, can be unset, I don't know what I was thinking... and what I meant by 

> Could trigger a false positive if we have multiple confirmations and one has the (invalid) empty string as 
> InResponseTo whereas a later one has a proper one ...

was that if an assertion has multiple subject confirmations (and I had set the `SC_VALID_IN_RESPONSE_TO`to the empty string), then the false positive would be a subject with 2 confirmations where ONE had the WRONG `InResponseTo` and one had the empty string as `InResponse` to. Such confirmation would falsely pass even though neither of the `InResponseTo` matches the request id (given that the request could not be unserialized so that `SC_VALID_IN_RESPONSE_TO` was set to the empty string). 

Since we can leave it unset, it should not be a problem. 

It seems it is enough with one confirmation passing the test on the subject (see `SAML20AssertionValidator::validateSubjectConfirmation`)

I've updated the code, let me know what you think!